### PR TITLE
bpo-41798: Allocate _decimal extension module CAPI on the heap

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue41798

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @shihai1991

File: Modules/_decimal/_decimal.c [hunk 1]
```diff
@@ -5716,8 +5734,13 @@ init_api(void)
     _decimal_api[PyDec_Get_INDEX] = (void *)PyDec_Get;
     _decimal_api[PyDec_GetConst_INDEX] = (void *)PyDec_GetConst;
 
-    return PyCapsule_New(_decimal_api, "_decimal._API", NULL);
+    PyObject *capsule = PyCapsule_New(_decimal_api, CAPSULE_NAME, destroy_api);
+    if (!capsule) {
+        free_api(_decimal_api);
```
Comment (shihai1991) [inline Modules/_decimal/_decimal.c]: It's fine to use `PyMem_Free()` directly in  here.

File: Modules/_decimal/_decimal.c [hunk 2]
```diff
@@ -5699,9 +5697,29 @@ PyDec_GetConst(const PyObject *v)
     return MPD(v);
 }
 
+static void 
+free_api(void *capi)
+{
+    PyMem_Free(capi);
+}
+
+#define CAPSULE_NAME "_decimal._API"
```
Comment (shihai1991) [inline Modules/_decimal/_decimal.c]: `_decimal._API` is used in https://github.com/python/cpython/blob/master/Include/pydecimal.h#L167, I suggest you can move this marco to the pydecimal.h.

File: Modules/_decimal/_decimal.c [hunk 3]
```diff
@@ -5699,9 +5697,29 @@ PyDec_GetConst(const PyObject *v)
...
+destroy_api(PyObject *obj)
+{
+    void *capi = PyCapsule_GetPointer(obj, CAPSULE_NAME);
+    free_api(capi);
+}
+
 static PyObject *
 init_api(void)
 {
+    void **_decimal_api = PyMem_Malloc(CPYTHON_DECIMAL_MAX_API * sizeof(void *));
```
Comment (shihai1991) [inline Modules/_decimal/_decimal.c]: using `PyMem_Calloc(size, sizeof(void *))` in here would be better? I am not sure.

Comment (erlend-aasland) [reply Modules/_decimal/_decimal.c]: Yes, it is definitely better. Thanks for pointing that out.

File: Modules/_decimal/_decimal.c (hunk 2 above)
Comment (erlend-aasland) [reply Modules/_decimal/_decimal.c]: I agree. I'll also rename it to `PyDec_CAPSULE_NAME` for consistence.

**Comment by erlend-aasland** [comment]
Thanks for your review, @shihai1991! PTAL

**Comment by vstinner** [comment]
Merged, thanks.

**Pull request merged**